sdk/go/auto/cmd.go

```diff
@@ -17,17 +17,203 @@ package auto
 import (
 	"bytes"
 	"context"
+	"fmt"
 	"io"
+	"net/http"
 	"os"
 	"os/exec"
+	"path"
+	"runtime"
+	"strings"
 
+	"github.com/blang/semver"
+	"github.com/pulumi/pulumi/sdk/v3"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/env"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/slice"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/httputil"
 )
 
 const unknownErrorCode = -2
 
-func runPulumiCommandSync(
-	ctx context.Context,
+// PulumiCommand manages the Pulumi CLI and runs operations.
+type PulumiCommand interface {
+	Run(ctx context.Context,
+		workdir string,
+		stdin io.Reader,
+		additionalOutput []io.Writer,
+		additionalErrorOutput []io.Writer,
+		additionalEnv []string,
+		args ...string) (string, string, int, error)
+	Version() semver.Version
+}
+
+type pulumiCommand struct {
+	version semver.Version
+	command string
+}
+
+// NewPulumiCommand creates a new PulumiCommand instance that uses a local CLI
+// installation. PulumiCommandOptions can be used to configure things like the
+// the installation root or to validate that the version meets a minimum
+// requirement. By default the pulumi binary found in $PATH is used.
+func NewPulumiCommand(opts ...PulumiCommandOption) (PulumiCommand, error) {
```

Comment:
We've been shying away from the go functional options pattern in new APIs, primarily because it's difficult to discover the options.

Another downside to the pattern is it "pollutes" the package with a bunch of functions for the various options, and these functions are only used if you're using `NewPulumiCommand` which is likely still going to be a niche thing to do for advanced scenarios. The original APIs that use the functional options pattern address the "pollution" problem by putting them in nested packages like `optup`, `optpreview`, etc.

Here are some more recent APIs we added that use an options struct:

https://github.com/pulumi/pulumi/blob/c71f001e39eb9f46276d73e385546cbe9c2e9d24/sdk/go/auto/workspace.go#L160-L164

https://github.com/pulumi/pulumi/blob/c71f001e39eb9f46276d73e385546cbe9c2e9d24/sdk/go/auto/workspace.go#L56-L59

https://github.com/pulumi/pulumi/blob/c71f001e39eb9f46276d73e385546cbe9c2e9d24/sdk/go/auto/workspace.go#L187-L190

https://github.com/pulumi/pulumi/blob/c71f001e39eb9f46276d73e385546cbe9c2e9d24/sdk/go/auto/workspace.go#L115-L118

Could we use a struct for this instead?

```go
func NewPulumiCommand(opts *PulumiCommandOptions) (PulumiCommand, error)

type PulumiCommandOptions struct {
	Version          semver.Version
	Root             string
	SkipVersionCheck bool
}
```

Reply:
Hah, I had it with a struct initially, but was mislead by LocalWorkspace :) 

I pushed a change to use structs again https://github.com/pulumi/pulumi/commit/7710a18cdec500782a8a36ddabedbd6ebeda0915